### PR TITLE
Allow the alertmanager in the garden ns to send email notifications

### PR DESCRIPTION
**Pull request by istvanballok**
**How to categorize this PR?**

/area monitoring
/kind regression

**What this PR does / why we need it**:

Allow the alertmanager in the garden namespace to send email notifications

The alertmanager in the garden namespace of a seed can be configured to send email notifications about the alerts of the shoot control plane Prometheus-es and the alerts of the aggregate Prometheus in the garden namespace.

See `charts/gardener/controlplane/values.yaml`, `global.alerting`.

We noticed it with quite some delay that probably since the recent improvements of the network policies, the `garden/alertmanager-0` pod does not have DNS and public/private network access and hence could not send out the email notifications.

```
k logs -n garden alertmanager-0 | tail -n 1

ts=2023-08-04T12:16:18.450Z
caller=dispatch.go:354
level=error
component=dispatcher
msg="Notify for alerts failed"
num_alerts=1
err="email-kubernetes-ops/email[0]:
     notify retry canceled after 2 attempts:
     establish connection to server:
     dial tcp 52.x.y.z:587: i/o timeout"
```

(DNS resolution works but the SMTP server can not be contacted)

or

```
err="establish connection to server: dial tcp: lookup email-smtp.foo.bar.com on 10.1.2.3:53:
read udp 10.4.5.6:42060->10.1.2.3:53:
i/o timeout"
```

(DNS resolution does not work)

This change grants DNS and public/private network access to the alertmanager in the garden namespace so that it can successfully send email notifications.

Note that the mail server could be either in a public or in a private network, hence both network paths: to the public and to the private network need to be allowed.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

cc @rfranzke @acumino @abdasgupta @rickardsjp 

**Release note**:

```other operator
gardenlet: A regression preventing the alertmanager in the garden namespace from sending email notifications is now fixed.
```

### CODE REVIEW

**Comment by istvanballok** [comment]
Could we cherry-pick this PR also for `v1.76.2`? cc @ialidzhikov

**Review by acumino — COMMENTED** [review]
/lgtm
/cc @ScheererJ

File: pkg/component/monitoring/charts/bootstrap/templates/alertmanager/alertmanager.yaml
Comment (abdasgupta) [inline pkg/component/monitoring/charts/bootstrap/templates/alertmanager/alertmanager.yaml:63]: Just a query actually, do we also need `networking.gardener.cloud/to-private-networks: allowed` here ? May be, somebody would like to setup their email server in a private network. I am not expert on how the network is setup in gardener. Feel free to ignore my comment if I am talking something absurd.

**Review by abdasgupta — CHANGES_REQUESTED** [review]

File: pkg/component/monitoring/charts/bootstrap/templates/alertmanager/alertmanager.yaml
Comment (istvanballok) [reply pkg/component/monitoring/charts/bootstrap/templates/alertmanager/alertmanager.yaml:63]: > do we also need `networking.gardener.cloud/to-private-networks: allowed` here?

Hi @abdasgupta, I also considered that but I could not think of a use case so I opted for skipping granting access to private networks in general. The SMTP server configuration is done on the Gardener landscape level (so it is not configurable e.g. for shoot owners) and we are using an SMTP server with a public IP, so I thought that public network access is sufficient for now. I think even in other Gardener landscapes, landscape operators probably use an SMTP server with a public IP because in the current setup the same IP has to be reachable from all the seeds.

With this reasoning, granting access to private networks is not required and I'm a bit unsure about the risk of granting general private network access if it is not necessary, so I'd proceed like this and looking for advice from @acumino for a second opinion. :)

Comment (abdasgupta) [reply pkg/component/monitoring/charts/bootstrap/templates/alertmanager/alertmanager.yaml:63]: Okay. I understood your reasoning @istvanballok . @acumino hope we are good. We can try to merge this PR then.

Comment (ScheererJ) [reply pkg/component/monitoring/charts/bootstrap/templates/alertmanager/alertmanager.yaml:63]: @istvanballok Please add `networking.gardener.cloud/to-private-networks: allowed` as well. There is no way of knowing if the mail server is in the private or public network. There might as well be Gardener installations that use a private mail server.

**Review by ScheererJ — CHANGES_REQUESTED** [review]
Nice finding. Please also add private network access as well.

Can you please also comment why this error message in the pull request description

```
     dial tcp 52.x.y.z:587: i/o timeout"
```

led you to add dns? I agree that it is most likely needed, but the error message rather suggests that name resolution already took place, which is surprising if DNS is blocked.

File: pkg/component/monitoring/charts/bootstrap/templates/alertmanager/alertmanager.yaml
Comment (istvanballok) [reply pkg/component/monitoring/charts/bootstrap/templates/alertmanager/alertmanager.yaml:63]: 👍 I amended the commit and the PR description.

**Review by acumino — COMMENTED** [review]
/lgtm

**Comment by istvanballok** [comment]
> Can you please also comment why this error message in the pull request description
>
>     dial tcp 52.x.y.z:587: i/o timeout"
>
> led you to add dns? I agree that it is most likely needed, but the error message rather suggests that name resolution already took place, which is surprising if DNS is blocked.

@ScheererJ: Actually I observed both kind of error messages: (I updated the PR description)

```
err="email-kubernetes-ops/email[0]:
     notify retry canceled after 2 attempts:
     establish connection to server:
     dial tcp 52.x.y.z:587: i/o timeout"
```

and

```
err="establish connection to server: dial tcp: lookup email-smtp.foo.bar.com on 10.1.2.3:53:
read udp 10.4.5.6:42060->10.1.2.3:53:
i/o timeout"
```

I observed the successful DNS resolution without the required network policy with an older version of Gardener: `v1.75`. I haven't figured out yet why the DNS resolution works in some of the cases even without the `to-dns: allowed` label. On the other hand, on the current version of Gardener that I verified in the local and test landscape setup, the label to allow dns access is clearly needed.

**Comment by acumino** [comment]
/approve
/hold 
@ScheererJ  Do you have any other concerns?

**Review by ScheererJ — APPROVED** [review]
Thanks for finding and fixing the issue.

/lgtm

**Pull request merged**
